server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java

```diff
@@ -277,4 +278,35 @@ public void testRemoteSegmentCleanup() throws Exception {
     public void testRemoteTranslogCleanup() throws Exception {
         verifyRemoteStoreCleanup(true);
     }
+
+    public void testStaleCommitDeletion() throws Exception {
+        internalCluster().startDataOnlyNodes(3);
+        createIndex(INDEX_NAME, remoteStoreIndexSettings(1, 10000));
+        int numberOfIterations = randomIntBetween(5, 15);
+        boolean invokeFlush = randomBoolean();
+        indexData(numberOfIterations, invokeFlush);
+        String indexUUID = client().admin()
+            .indices()
+            .prepareGetSettings(INDEX_NAME)
+            .get()
+            .getSetting(INDEX_NAME, IndexMetadata.SETTING_INDEX_UUID);
+        Path indexPath = Path.of(String.valueOf(absolutePath), indexUUID, "/0/segments/metadata");
+        if (invokeFlush) {
+            // Delete is async.
+            assertBusy(() -> {
+                try {
+                    int actualFileCount = getFileCount(indexPath);
+                    if (numberOfIterations <= RemoteStoreRefreshListener.LAST_N_METADATA_FILES_TO_KEEP) {
+                        assertEquals(numberOfIterations, actualFileCount);
+                    } else {
+                        // As delete is async its possible that the file gets created before the deletion or after
+                        // deletion.
+                        assertTrue(actualFileCount >= 10 || actualFileCount <= 11);
```

Comment:
As we make sure only one thread is executing `deleteStaleCommits` at any given time, is it possible to have the metadata file count more than 11? Assume that threads of `REMOTE_PURGE` are busy and delete stale commits thread is in queue. In the meantime, another commit happens and skip the delete part as semaphore is already acquired.